src/kernel/src/obj/copy.rs

```diff
@@ -24,86 +27,489 @@ fn split_range(
+/// remapped appropriately for any mapping of the objects. This ensures that the source object is
+/// "checkpointed" before copying, and that the destination object cannot be read in the region being
+/// overwritten until the copy is done.
 pub fn copy_ranges(
     src: &ObjectRef,
-    src_start: PageNumber,
+    src_off: usize,
     dest: &ObjectRef,
-    dest_start: PageNumber,
-    length: usize,
+    dest_off: usize,
+    byte_length: usize,
 ) {
-    let (src_tree, mut dest_tree) = crate::utils::lock_two(&src.range_tree, &dest.range_tree);
+    let src_start = PageNumber::from_offset(src_off);
+    let dest_start = PageNumber::from_offset(dest_off);
+
+    let start_offset = src_off % PageNumber::PAGE_SIZE;
+    let end_offset = (src_off + byte_length) % PageNumber::PAGE_SIZE;
+
+    // Number of pages that will be touched, including partial pages.
+    // By subtracting the partial pages, we are left with the full pages,
+    // and then we can add in how many partial pages we'll be copying.
+    let nr_pages: usize = byte_length.saturating_sub(start_offset + end_offset)
+        / PageNumber::PAGE_SIZE
+        + match (start_offset, end_offset) {
+            (0, 0) => 0,
+            (0, _) | (_, 0) => 1,
+            (_, _) => 2,
+        };
+    // Step 1: lock the page trees for the objects, in a canonical order.
+    let (mut src_tree, mut dest_tree) = crate::utils::lock_two(&src.range_tree, &dest.range_tree);
+
+    // Step 2: Invalidate the page ranges. In the destination, we fully unmap the object for that range. In the source,
+    // we only need to ensure that no one modifies pages, so we just write-protect it.
+    src.invalidate(
+        src_start..src_start.offset(nr_pages),
+        InvalidateMode::WriteProtect,
+    );
+    dest.invalidate(
+        dest_start..dest_start.offset(nr_pages),
+        InvalidateMode::Full,
+    );
 
+    // if we can't do COW copy, then fallback to full copy.
+    if src_off % PageNumber::PAGE_SIZE != dest_off % PageNumber::PAGE_SIZE {
+        copy_bytes(
+            &mut src_tree,
+            src_off,
+            &mut dest_tree,
+            dest_off,
+            byte_length,
+        );
+        return;
+    }
+
+    // Step 3a: Copy any non-full-page at the start
     let mut dest_point = dest_start;
     let mut src_point = src_start;
-    let mut rem = length;
-    let ranges = src_tree.range(src_start..src_start.offset(length));
-    for range in ranges {
-        if src_point < *range.0 {
-            /* TODO: we'll need to ensure all backing pages are present if we get here */
-            let diff = *range.0 - src_point;
-            dest_point = dest_point.offset(diff);
-            rem -= diff;
+    let mut remaining_pages = nr_pages;
+    if start_offset != 0 {
+        copy_single(
+            &mut dest_tree,
+            &mut src_tree,
+            dest_point,
+            src_point,
+            start_offset,
+            PageNumber::PAGE_SIZE,
```

Comment:
Isn't this suppose to copy a partial page?

Reply:
It does, copying from the start offset to the end of the page (last two args).